src/file_operations.js

```diff
@@ -5,74 +5,113 @@ const util = require("util")
 const path = require("path")
 const cliProgress = require("cli-progress")
 
+const {
+  MISSING_COURSE_ERROR_MESSAGE,
+  NO_COURSES_FOUND_MESSAGE
+} = require("./constants")
 const { directoryExists } = require("./helpers")
 const markdownGenerators = require("./markdown_generators")
 const loggers = require("./loggers")
+const helpers = require("./helpers")
 
 const progressBar = new cliProgress.SingleBar(
   { stopOnComplete: true },
   cliProgress.Presets.shades_classic
 )
 const readdir = util.promisify(fs.readdir)
 
-const scanCourses = (source, destination) => {
+const scanCourses = (inputPath, outputPath, jsonPath = null) => {
   /*
-    This function scans the source directory for course folders
+    This function scans the input directory for course folders
   */
-  // Make sure that the source and destination arguments have been passed and they are directories
-  if (!directoryExists(source)) {
-    throw new Error("Invalid source directory")
+  // Make sure that the input and output arguments have been passed and they are directories
+  if (!directoryExists(inputPath)) {
+    throw new Error("Invalid input directory")
   }
-  if (!directoryExists(destination)) {
-    throw new Error("Invalid destination directory")
+  if (!directoryExists(outputPath)) {
+    throw new Error("Invalid output directory")
   }
-  // Iterate all subdirectories under source
-  const contents = fs.readdirSync(source)
-  const totalDirectories = contents.filter(file =>
-    directoryExists(path.join(source, file))
-  ).length
-  console.log(`Converting ${totalDirectories} courses to Hugo markdown...`)
-  progressBar.start(totalDirectories, 0)
-  contents.forEach(file => {
-    const coursePath = path.join(source, file)
-    if (fs.lstatSync(coursePath).isDirectory()) {
-      // If the item is indeed a directory, read all files in it
-      scanCourse(coursePath, destination).then(() => {
-        progressBar.increment()
-      })
+  if (jsonPath) {
+    // Iterate only the courses specified in the coursesJson
+    const coursesJson = JSON.parse(fs.readFileSync(jsonPath))
+    const totalCourses = coursesJson["courses"].length
+    if (totalCourses > 0) {
+      console.log(`Converting ${totalCourses} courses to Hugo markdown...`)
+      progressBar.start(totalCourses, 0)
+      coursesJson["courses"].forEach(course =>
+        scanCourse(inputPath, outputPath, course).then(() => {
+          progressBar.increment()
+        })
+      )
+    } else {
+      console.log(NO_COURSES_FOUND_MESSAGE)
     }
-  })
+  } else {
```

Comment:
the contents of this `else` look pretty close to the body of the `if (jsonPath)` - maybe we could use the conditional just to get the course list, and then use the same logic to go through it?

something like

```js
const courseList = jsonPath
  ? JSON.parse(fs.readFileSync(jsonPath))["courses"]
  : fs.readDirSync(inputPath)

const numCourses = jsonPath
  ? courseList.length
  : courseList.filter(file =>
      directoryExists(path.join(inputPath, file))
    ).length
```

then I think you could from that point on treat them the same, if I'm reading things correctly

Reply:
A very good point!  What you posted worked great.